src/databricks/labs/ucx/source_code/graph.py

```diff
@@ -480,3 +506,49 @@ class MaybeGraph:
     @property
     def failed(self):
         return len(self.problems) > 0
+
+
+T = TypeVar("T")
+
+# non-top-level import is required to avoid cyclic dependency
+# pylint: disable=wrong-import-position, cyclic-import
+from databricks.labs.ucx.source_code.notebooks.commands import PipCommand  # noqa: E402
+
+
+class MagicCommand(NodeBase):
+
+    @classmethod
+    def extract_from_tree(
+        cls, tree: Tree, problem_factory: ProblemFactory
+    ) -> tuple[list[MagicCommand], list[DependencyProblem]]:
+        problems: list[DependencyProblem] = []
+        commands: list[MagicCommand] = []
+        try:
+            nodes = tree.locate(Call, [("magic_command", Name)])
+            for command in cls._make_commands_for_magic_command_call_nodes(nodes):
+                commands.append(command)
+            return commands, problems
+        except Exception as e:  # pylint: disable=broad-except
+            problem = problem_factory('internal-error', f"While checking magic commands: {e}", tree.root)
+            problems.append(problem)
+            return [], problems
+
+    @classmethod
+    def _make_commands_for_magic_command_call_nodes(cls, nodes: list[Call]):
+        for node in nodes:
+            arg = node.args[0]
+            if isinstance(arg, Const):
+                yield MagicCommand(node, arg.value)
+
+    def __init__(self, node: NodeNG, command: str):
+        super().__init__(node)
+        data = base64.b64decode(command)
+        self._command = data.decode("ascii")
+
+    def execute(self, return_type: T, **kwargs) -> T:
```

Comment:
if you move it to commands.py, then you don't need to add generic type here.

Reply:
Dropping for now but I suspect it will come back when we execute from a linter, which will require a different return type.